File: DynamicKey/AgoraDynamicKey/dart/test/convoai_token_builder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:agora_token_generator/agora_token_generator.dart';

void main() {
  const appId = '970CA35de60c44645bbae8a215061b33';
  const appCertificate = '5CFd2fd1755d40ecb72977518be15d3b';
  const channelName = '7d72365eb983485397e3e3f9d460bdda';
  const rtcAccount = '2882341273';
  const rtmUserId = '2882341273';
  const rtcTokenExpireSeconds = 600;
  const joinChannelPrivilegeExpire = 600;
  const publishAudioPrivilegeExpire = 600;
  const publishVideoPrivilegeExpire = 600;
  const publishDataPrivilegeExpire = 600;
  const rtmTokenExpireSeconds = 600;

  test('builds a publisher ConvoAI token', () {
    final token = ConvoAITokenBuilder.buildToken(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      rtcAccount: rtcAccount,
      rtcRole: RtcRole.publisher,
      rtcTokenExpireSeconds: rtcTokenExpireSeconds,
      joinChannelPrivilegeExpire: joinChannelPrivilegeExpire,
      publishAudioPrivilegeExpire: publishAudioPrivilegeExpire,
      publishVideoPrivilegeExpire: publishVideoPrivilegeExpire,
      publishDataPrivilegeExpire: publishDataPrivilegeExpire,
      rtmUserId: rtmUserId,
      rtmTokenExpireSeconds: rtmTokenExpireSeconds,
    );

    final parsed = AccessToken.empty();
    expect(parsed.parse(token), isTrue);
    final rtc = parsed.getServices(Service.RTC).single as ServiceRtc;
    final rtm = parsed.getServices(Service.RTM).single as ServiceRtm;
    final convoai = parsed.getServices(Service.CONVOAI).single as ServiceConvoAI;
    expect(parsed.appId, appId);
    expect(parsed.expire, rtcTokenExpireSeconds);
    expect(rtc.channelName, channelName);
    expect(rtc.uid, rtcAccount);
    expect(rtc.privileges, {
      ServiceRtc.privilegeJoinChannel: joinChannelPrivilegeExpire,
      ServiceRtc.privilegePublishAudioStream: publishAudioPrivilegeExpire,
      ServiceRtc.privilegePublishVideoStream: publishVideoPrivilegeExpire,
      ServiceRtc.privilegePublishDataStream: publishDataPrivilegeExpire,
    });
    expect(rtm.userId, rtmUserId);
    expect(rtm.privileges[ServiceRtm.privilegeLogin], rtmTokenExpireSeconds);
    expect(convoai.privileges, isEmpty);
  });

  test('builds a subscriber ConvoAI token', () {
    final token = ConvoAITokenBuilder.buildToken(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      rtcAccount: rtcAccount,
      rtcRole: RtcRole.subscriber,
      rtcTokenExpireSeconds: rtcTokenExpireSeconds,
      joinChannelPrivilegeExpire: joinChannelPrivilegeExpire,
      publishAudioPrivilegeExpire: publishAudioPrivilegeExpire,
      publishVideoPrivilegeExpire: publishVideoPrivilegeExpire,
      publishDataPrivilegeExpire: publishDataPrivilegeExpire,
      rtmUserId: rtmUserId,
      rtmTokenExpireSeconds: rtmTokenExpireSeconds,
    );

    final parsed = AccessToken.empty();
    expect(parsed.parse(token), isTrue);
    final rtc = parsed.getServices(Service.RTC).single as ServiceRtc;
    expect(rtc.privileges, {
      ServiceRtc.privilegeJoinChannel: joinChannelPrivilegeExpire,
    });
  });
}

File: DynamicKey/AgoraDynamicKey/deno/test/ConvoAITokenBuilderTest.js
import { ConvoAITokenBuilder } from '../src/ConvoAITokenBuilder.js'
import { AccessToken2, kConvoAIServiceType, kRtcServiceType, kRtmServiceType, ServiceRtc, ServiceRtm } from '../src/AccessToken2.js'
import { Role } from '../src/RtcTokenBuilder2.js'
import { assert, assertEquals } from 'https://deno.land/std/testing/asserts.ts'

const appId = '970CA35de60c44645bbae8a215061b33'
const appCertificate = '5CFd2fd1755d40ecb72977518be15d3b'
const channelName = '7d72365eb983485397e3e3f9d460bdda'
const rtcAccount = '2882341273'
const rtmUserId = '2882341273'
const rtcTokenExpire = 600
const joinChannelPrivilegeExpire = 600
const pubAudioPrivilegeExpire = 600
const pubVideoPrivilegeExpire = 600
const pubDataStreamPrivilegeExpire = 600
const rtmTokenExpire = 600

Deno.test('ConvoAITokenBuilder publisher token', () => {
  const token = ConvoAITokenBuilder.buildToken(
    appId, appCertificate, channelName, rtcAccount, Role.PUBLISHER, rtcTokenExpire,
    joinChannelPrivilegeExpire, pubAudioPrivilegeExpire, pubVideoPrivilegeExpire,
    pubDataStreamPrivilegeExpire, rtmUserId, rtmTokenExpire,
  )
  const parsed = new AccessToken2('', '', 0, 0)
  assert(parsed.from_string(token))
  const rtc = parsed.getServices(kRtcServiceType)[0]
  const rtm = parsed.getServices(kRtmServiceType)[0]
  const convoai = parsed.getServices(kConvoAIServiceType)[0]
  assertEquals(parsed.appId, appId)
  assertEquals(parsed.expire, rtcTokenExpire)
  assertEquals(rtc.__channel_name, channelName)
  assertEquals(rtc.__uid, rtcAccount)
  assertEquals(rtc.__privileges[ServiceRtc.kPrivilegeJoinChannel], joinChannelPrivilegeExpire)
  assertEquals(rtc.__privileges[ServiceRtc.kPrivilegePublishAudioStream], pubAudioPrivilegeExpire)
  assertEquals(rtc.__privileges[ServiceRtc.kPrivilegePublishVideoStream], pubVideoPrivilegeExpire)
  assertEquals(rtc.__privileges[ServiceRtc.kPrivilegePublishDataStream], pubDataStreamPrivilegeExpire)
  assertEquals(rtm.__user_id, rtmUserId)
  assertEquals(rtm.__privileges[ServiceRtm.kPrivilegeLogin], rtmTokenExpire)
  assertEquals(convoai.__privileges, {})
})

Deno.test('ConvoAITokenBuilder subscriber token', () => {
  const token = ConvoAITokenBuilder.buildToken(
    appId, appCertificate, channelName, rtcAccount, Role.SUBSCRIBER, rtcTokenExpire,
    joinChannelPrivilegeExpire, pubAudioPrivilegeExpire, pubVideoPrivilegeExpire,
    pubDataStreamPrivilegeExpire, rtmUserId, rtmTokenExpire,
  )
  const parsed = new AccessToken2('', '', 0, 0)
  assert(parsed.from_string(token))
  const rtc = parsed.getServices(kRtcServiceType)[0]
  assertEquals(rtc.__privileges, { [ServiceRtc.kPrivilegeJoinChannel]: joinChannelPrivilegeExpire })
})
